### PR TITLE
Upstream merge 

### PR DESCRIPTION
**Pull request by djramic**
There are still 6 tests that need to be fixed:
 -Conversion/EmulateFp8ExtTrunc/Integration/roundtrip.mlir
 -fusion/pr-e2e/migraphx-group-conv2d.e2e.mlir
 -rocmlir-driver/check-residency.mlir
 -xmir/pr-e2e/resnet50_blk2.mlir
 -xmir/resnet50_blk.async.mlir
 -xmir/mobilenetv1.mid.mlir

### CODE REVIEW

**Comment by krzysz00** [comment]
Unless this is very blocking, I'd like to review this once I'm back later this week.
Or, in the alternative, I'd like to be able to check the changes that got made after the fact

File: external/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp [hunk 1]
```diff
@@ -294,7 +294,7 @@ calculateVectorIndex(Value *Ptr,
 
   auto I = GEPIdx.find(GEP);
   assert(I != GEPIdx.end() && "Must have entry for GEP!");
-   return isa<Constant>(I->second) ? I->second :  ConstantInt::getNullValue(Type::getInt32Ty(Ptr->getContext()));
+   return I->second;
```
Comment (sjw36) [inline external/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp]: fix spacing. Probably just need to run clang-format

**Review by sjw36 — APPROVED** [review]

**Review by sjw36 — CHANGES_REQUESTED** [review]
Oops, didn't mean to approve the whole upstream merge.
We still need some perf analysis and a few more things fixed.

**Comment by krzysz00** [comment]
Notes from my staring at the diffs
1. In `ROCDLOps.td`, there's a lot of diff noise from things like `ROCDL_CvtF32Fp8Op`. I suspect most of it is that things got shuffled around, but I do remember those operations getting renamed. Please ensure that that file is a superset of upstream (that is, we can add stuff, but we shouldn't remove it)
2. In `external/llvm-project/mlir/include/mlir/Dialect/Arith/Transforms/Passes.h`, we've got duplicate declarations of `populateEmulateUnsupportedFloatsConversions` and friends. Please remove the ones that aren't upstream
3. There seems to be diff noise in `def Tosa_CastOp:` - is that trailing whitespace being stripped? We might want to clean that up to match upstream.
4. `external/llvm-project/mlir/include/mlir/IR/Value.h` has grown an extra `const` on `replaceAllUsesWith` 
5. Remove `getMaybeVectorConstant`, from `external/llvm-project/mlir/lib/Conversion/ArithToAMDGPU/ArithToAMDGPU.cpp` - it got killed in upstream code review
6. `external/llvm-project/mlir/lib/Conversion/ControlFlowToLLVM/CMakeLists.tx`: spare `MLIRTransforms` in the dependency list. Also `external/llvm-project/mlir/lib/Conversion/FuncToLLVM/CMakeLists.txt` . Also `external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/CMakeLists.txt` 
7. In `configureGpuToROCDLConversionLegality` in `external/llvm-project/mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp `, you can swap `::mlir;:ROCDL` for `ROCDL` to match upstream
8. Same file, remove
    ```
    +#include "mlir/Dialect/LLVMIR/Transforms/Passes.h"
    +
    ```
    and
    ```
    #include "mlir/Pass/PassManager.h"
    ```
    as redundant
9. Missing newline at end of `TosaToLinalgNamed.cpp` (and we've got a followup item for Yi to update her upstream PR)
10. `TosaToLinalgPass.cpp` lists `target.addLegalOp<tosa::ReshapeOp>();` twice - remove the earlier one
11. In `GPUDialect.cpp`, prefer upstream's `void GPUFuncOp::setWorkgroupAttributionAttr` (might just be a location change)
12. In `LLVMIR/IR/LLVMDialect.cpp`, fix trivial extra whitespace
13. In `external/llvm-project/mlir/lib/Dialect/LLVMIR/Transforms/CMakeLists.txt` , no-longer-needed dependency on `MLIRLLVMCommonConversion`
14. In `external/llvm-project/mlir/lib/Dialect/Math/Transforms/LegalizeToF32.cpp`, just copy in upstream's version (which'll also fix the lack of trailing newline)
15. `Tosa/IR/TosaOps.cpp`:  blank line before in `buildConvOpWithQuantInfo` before `addOperands` removed pointlessly
16. Copy ` external/llvm-project/mlir/test/Conversion/TosaToTensor/tosa-to-tensor.mlir` from upstream to fix trivial (they have `%.+`, we had `%.*`) discrepancies
17. Fix `external/llvm-project/mlir/test/Conversion/TosaToLinalg/tosa-to-linalg.mlir ` - we have hardcoded `%arg0` in some places where upstream had `%[[ARG0]]`
18. Remove redundant copy of `llvm.func @rocdl_8bit_floats` in `external/llvm-project/mlir/test/Dialect/LLVMIR/rocdl.mlir`
19. Try restoring `external/llvm-project/mlir/test/Integration/GPU/ROCM/printf.mlir` now that I've fixed the device library segfaults
20. Remove `void registerTestTilingInterface();`  from `external/llvm-project/mlir/tools/mlir-opt/mlir-opt.cpp ` - it's a function declaration that doesn't correspond to a function anymore

Changes with rocmlir
1. In `mlir/lib/Conversion/EmulateFp8ExtTrunc/EmulateFp8ExtTrunc.cpp`, you removed an `&rewriter` when creating an OpBuilder. The reason that was passed in was to make sure to preserve the debug listener. See if there's a new API for that.
2. In `mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp `
    ```
    +  auto filtered_range =  op->getDiscardableAttrDictionary();
    +  std::vector<NamedAttribute> filtered_attrs(filtered_range.begin(), filtered_range.end());
    ```
    should be something like `SmallVector<NamedAttribute> filteredAttrs = llvm::to_vector(op->getDiscardableAttrDictionary());` . That saves us a heap allocation in the common case.
3. Same file, in the quantize linear lowering:
    ```
    +    FloatAttr minFatt = rewriter.getF32FloatAttr(minF.convertToDouble());
   +    FloatAttr maxFatt = rewriter.getF32FloatAttr(maxF.convertToDouble());
   ```
    is wrong - if those need to be attributes, they should be attributes of the same type as `minF` and `maxF`.
5. When you were hacking on the `notify` methods in `AlignTiling.cpp`, please look at the corresponding methods in other `Listener` implementtaions and modify ours accordingly instead of just ignoring the extra arguments
6. Why `MLIRUBToLLVM` and `MLIRComplexToLLVM` in `mlir/lib/Dialect/Rock/Transforms/CMakeLists.txt `? We don't use those. On top of that, those should be added to `external/mlir-hal/lib/Dialect/MHAL/Pipelines/CMakeLists.txt` , because the Rock transforms don't depend on LLVM lowerings
7. You shouldn't be making `MLIRAffineTransformOps` a dependency of `rocmlir-driver` - please find what actually depends on it.
8. Why did you add
    ```
    +get_property(mlir_dialect_libs GLOBAL PROPERTY MLIR_DIALECT_LIBS)
    +get_property(mlir_conversion_libs GLOBAL PROPERTY MLIR_CONVERSION_LIBS)
    ```
    to `rocmlir-driver`'s CMakeLists.txt? It doesn't do anything.
9. Spare newline in CMake dependency list in `mlir/lib/Dialect/Rock/IR/CMakeLists.txt`
10. Similarly, why does `MLIRRockOps` now depend on `MLIRLinalgTransformOps`?
11. Could I get some info on the change to `mlir/test/fusion/tosa-tp-highlevel.mlir`? It'll be good to make sure we haven't had any big side effects from whatever changed upstream

Overall, it looks like we didn't have to do a whole lot for this merge, but there's always some random noise.

(Also, for future us https://github.com/llvm/llvm-project/pull/67301 is the cherry-picked PR that adds the FMINIMUM/FMAXIMUM stuff to our copy of LLVM)

**Comment by krzysz00** [comment]
Re the dependency removal, I'm assuming that everything still builds with that change.

I suspect that, longer-term, we'll want to slide the host code conversion over to the convert-to-llvm interface.

**Comment by sjw36** [comment]
Looks like Extensions are still in development. We should probably give it a bit to settle.

**Comment by krzysz00** [comment]
In further review
- I think most of the `external/` changes are addressed, I don't see anything worth nitpicking over further
- `QuantizeLinearConverter` in MIGraphxToTosa.cpp still loses type information with `convertToDouble()` - please revert that change
- I don't think `MLIRLinalgTransformOps` needs to be dependcy of `MLIRRockOps`? It might need to go on the `Transforms/CMakeLists.txt` (or on `Utils`)

The one I'm actually worried about is `QuantizeLinear` there. Please fix that.

**Comment by djramic** [comment]
> In further review
> 
>     * I think most of the `external/` changes are addressed, I don't see anything worth nitpicking over further
> 
>     * `QuantizeLinearConverter` in MIGraphxToTosa.cpp still loses type information with `convertToDouble()` - please revert that change
> 
>     * I don't think `MLIRLinalgTransformOps` needs to be dependcy of `MLIRRockOps`? It might need to go on the `Transforms/CMakeLists.txt` (or on `Utils`)
> 
> 
> The one I'm actually worried about is `QuantizeLinear` there. Please fix that.

-I've added **MLIRLinalgTransformOps** to **MLIRRockOps** to resolve the linking error for **libMLIRRockOps**, related to mlir::linalg::GenericOp. Do you have a suggestion on where to move it? If i put in in Transforms/CMakeLists.txt, I still have the error.
-In QuantizeLinearConverter, can I use this conversion: **rewriter.getFloatAttr(rewriter.getF32Type(), minF);**

**Comment by sjw36** [comment]
> * `QuantizeLinearConverter` in MIGraphxToTosa.cpp still loses type information with `convertToDouble()` - please revert that change

@krzysz00 the previous implementation was F32Attr, so current change does what it used to do. 
It appears Fabrizio changed to allow support for fp16/bf16 also.

Author: fabrizio-indirli <fabrizio.indirli@arm.com>
Date:   Fri Oct 20 09:44:01 2023 +0100

```
    [mlir][Tosa] fix fp16/bf16 support for Clamp min/max attributes (#69192)
    
    In TOSA MLIR dialect, fix the definition of the Clamp op to
    accept fp16 & bf16 datatype for the min_fp and max_fp attributes.
    Add ClampOp verifier to check attributes types compatibility.
    Add related test cases in Tosa/ops.mlir.

```

**Comment by djramic** [comment]
@giuseros has approved the performance.

Here are the charts showing the performance differences:
[MLIR Performance Changes, Conv_ Chip = gfx90a.pdf](https://github.com/ROCm/rocMLIR/files/14640824/MLIR.Performance.Changes.Conv_.Chip.gfx90a.pdf)
[MLIR Performance Changes, GEMM_ Chip = gfx90a.pdf](https://github.com/ROCm/rocMLIR/files/14640837/MLIR.Performance.Changes.GEMM_.Chip.gfx90a.pdf)
[MLIR Performance Changes, Conv_ Chip = gfx908.pdf](https://github.com/ROCm/rocMLIR/files/14640842/MLIR.Performance.Changes.Conv_.Chip.gfx908.pdf)
[MLIR Performance Changes, GEMM_ Chip = gfx908.pdf](https://github.com/ROCm/rocMLIR/files/14640849/MLIR.Performance.Changes.GEMM_.Chip.gfx908.pdf)
[MLIR Performance Changes, Conv_ Chip = gfx1100.pdf](https://github.com/ROCm/rocMLIR/files/14640853/MLIR.Performance.Changes.Conv_.Chip.gfx1100.pdf)
[MLIR Performance Changes, GEMM_ Chip = gfx1100.pdf](https://github.com/ROCm/rocMLIR/files/14640857/MLIR.Performance.Changes.GEMM_.Chip.gfx1100.pdf)
[MLIR Performance Changes, Conv_ Chip = gfx1030.pdf](https://github.com/ROCm/rocMLIR/files/14640858/MLIR.Performance.Changes.Conv_.Chip.gfx1030.pdf)
[MLIR Performance Changes, GEMM_ Chip = gfx1030.pdf](https://github.com/ROCm/rocMLIR/files/14640862/MLIR.Performance.Changes.GEMM_.Chip.gfx1030.pdf)

@krzysz00 Can you review the latest commit?

**Comment by jerryyin** [comment]
@sjw36 Your review is blocking this PR. Are there additional concerns you have?

**Comment by sjw36** [comment]
No additional concerns. Un-blocking..

**Review by sjw36 — APPROVED** [review]

**Pull request merged**